test/python/test_pool.py

```diff
@@ -0,0 +1,164 @@
+
+import pytest
+from pytest_bdd import given, scenario, then, when, parsers
+from common.command import run_cmd
+import mayastor_pb2 as pb
+
+
+@scenario('features/pool.feature', 'creating a pool using disk with invalid block size')
+def test_creating_a_pool_using_disk_with_invalid_block_size():
+    'Creating a pool using disk with invalid block size.'
+
+
+@scenario('features/pool.feature', 'creating a pool with a name that already exists')
+def test_creating_a_pool_with_a_name_that_already_exists():
+    'Creating a pool with a name that already exists.'
+
+
+@scenario('features/pool.feature', 'creating a pool with an AIO disk')
+def test_creating_a_pool_with_an_aio_disk():
+    'Creating a pool with an AIO disk.'
+
+
+@scenario('features/pool.feature', 'creating a pool with multiple disks')
+def test_creating_a_pool_with_multiple_disks():
+    'Creating a pool with multiple disks.'
+
+
+@scenario('features/pool.feature', 'destroying a pool')
+def test_destroying_a_pool():
+    'Destroying a pool.'
+
+
+@scenario('features/pool.feature', 'destroying a pool that does not exist')
+def test_destroying_a_pool_that_does_not_exist():
+    'Destroying a pool that does not exist.'
+
+
+@scenario('features/pool.feature', 'listing pools')
+def test_listing_pools():
+    'Listing pools.'
+
+
+@pytest.fixture(scope="module")
+def mayastor_instance(mayastor_mod):
+    yield mayastor_mod['ms0']
+
+
+@pytest.fixture(scope="module")
+def find_pool(mayastor_instance):
+    def lookup(name):
+        for pool in mayastor_instance.ms.ListPools(pb.Null()).pools:
+            if pool.name == name:
+                return pool
+        return None
+    yield lookup
+
+
+@pytest.fixture
+def create_image_file():
+    filename = '/tmp/ms0-aio.img'
+    run_cmd(f"rm -f '{filename}'", True)
+    run_cmd(f"truncate -s 64M '{filename}'", True)
+    yield filename
+    run_cmd(f"rm -f '{filename}'", True)
+
+
+@pytest.fixture
+def mayastor_state(mayastor_instance):
+    state = {}
+    yield state
+    if 'pool' in state:
+        mayastor_instance.ms.DestroyPool(pb.DestroyPoolRequest(name=state['pool'].name))
+
+
+@given(parsers.parse('a mayastor instance "{name}"'))
+def get_instance(mayastor_instance):
+    pass
+
+@given('a pool')
+def create_pool(mayastor_instance, mayastor_state):
+    mayastor_state['pool'] = mayastor_instance.ms.CreatePool(pb.CreatePoolRequest(name='p0', disks=['malloc:///disk0?size_mb=100']))
+
+
+@when('the user attempts to create a pool specifying a URI representing an aio disk')
+def create_pool_from_aio_disk(mayastor_instance, create_image_file, mayastor_state):
+    try:
+        mayastor_state['pool'] = mayastor_instance.ms.CreatePool(pb.CreatePoolRequest(name='p0', disks=[f"aio://{create_image_file}"]))
+    except Exception as error:
+        mayastor_state['error'] = error
```

Comment:
I see several of these, why not raise the exception? (i.e test failed).

Reply:
Some (although by no means all) tests require the gRPC command to fail so, so raising (or rather not catching) the exception is not always appropriate. The real reason however is an attempt to adhere to the BDD flow: where a 'given' clause sets up the environment, the 'when' clause performs the action under test, and it is the 'then' clause that is responsible for validating the behaviour.